block/azure/adapter.go

```diff
@@ -0,0 +1,505 @@
+	if opts.StorageClass != nil {
+		res.BlobAccessTier = azblob.AccessTierType(*opts.StorageClass)
+	}
+	return res
+}
+
+func (a *Adapter) Put(obj block.ObjectPointer, sizeBytes int64, reader io.Reader, opts block.PutOpts) error {
+	var err error
+	defer reportMetrics("Put", time.Now(), &sizeBytes, &err)
+
+	qualifiedKey, err := resolveNamespace(obj)
+	if err != nil {
+		return err
+	}
+	blobURL := a.getBlockBlobURL(qualifiedKey.StorageNamespace, qualifiedKey.Key)
+
+	_, err = azblob.UploadStreamToBlockBlob(a.ctx, reader, blobURL, translatePutOpts(opts))
+	return err
+}
+
+func (a *Adapter) Get(obj block.ObjectPointer, _ int64) (io.ReadCloser, error) {
+	var err error
+	defer reportMetrics("Get", time.Now(), nil, &err)
+
+	return a.Download(obj, 0, azblob.CountToEnd)
+}
+
+func (a *Adapter) GetRange(obj block.ObjectPointer, startPosition int64, endPosition int64) (io.ReadCloser, error) {
+	var err error
+	defer reportMetrics("GetRange", time.Now(), nil, &err)
+
+	return a.Download(obj, startPosition, endPosition-startPosition+1)
+}
+
+func (a *Adapter) Download(obj block.ObjectPointer, offset, count int64) (io.ReadCloser, error) {
+	qualifiedKey, err := resolveNamespace(obj)
+	if err != nil {
+		return nil, err
+	}
+	blobURL := a.getBlobURL(qualifiedKey.StorageNamespace, qualifiedKey.Key)
+
+	keyOptions := azblob.ClientProvidedKeyOptions{}
+	downloadResponse, err := blobURL.Download(a.ctx, offset, count, azblob.BlobAccessConditions{}, false, keyOptions)
+
+	if err != nil {
+		return nil, err
+	}
+	// NOTE: automatically retries are performed if the connection fails
+	bodyStream := downloadResponse.Body(a.configurations.retryReaderOptions)
+	return bodyStream, nil
+}
+
+func (a *Adapter) Walk(walkOpt block.WalkOpts, walkFn block.WalkFunc) error {
+	var err error
+	defer reportMetrics("Walk", time.Now(), nil, &err)
+
+	qualifiedPrefix, err := resolveNamespacePrefix(walkOpt)
+	if err != nil {
+		return err
+	}
+
+	containerURL := a.getContainerURL(qualifiedPrefix.StorageNamespace)
+
+	for marker := (azblob.Marker{}); marker.NotDone(); {
+		listBlob, err := containerURL.ListBlobsFlatSegment(a.ctx, marker, azblob.ListBlobsSegmentOptions{Prefix: qualifiedPrefix.Prefix})
+		if err != nil {
+			return err
+		}
+
+		marker = listBlob.NextMarker
+		for _, blobInfo := range listBlob.Segment.BlobItems {
+			if err := walkFn(blobInfo.Name); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (a *Adapter) Exists(obj block.ObjectPointer) (bool, error) {
+	var err error
+	defer reportMetrics("Exists", time.Now(), nil, &err)
+
+	qualifiedKey, err := resolveNamespace(obj)
+	if err != nil {
+		return false, err
+	}
+
+	blobURL := a.getBlobURL(qualifiedKey.StorageNamespace, qualifiedKey.Key)
+	_, err = blobURL.GetProperties(a.ctx, azblob.BlobAccessConditions{}, azblob.ClientProvidedKeyOptions{})
+	if err != nil {
+		if err.(azblob.StorageError).ServiceCode() == azblob.ServiceCodeBlobNotFound {
+			return false, nil
+		}
+		return false, err
+	}
```

Comment:
This one should probably do the same using Go's error package
```suggestion
	var storageErr azblob.StorageError
	if errors.As(err, &storageErr)  && storageErr.ServiceCode() == azblob.ServiceCodeBlobNotFound {
		return false, nil
	} else if err != nil {
		return false, err
	}
```